pkg/rhcos/release/release.go

```diff
@@ -0,0 +1,180 @@
+	"github.com/containers/image/pkg/blobinfocache"
+	"github.com/containers/image/signature"
+	"github.com/containers/image/transports"
+	"github.com/containers/image/types"
+	"github.com/docker/distribution/reference"
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+// RHCOSBuild extracts the RHCOS build version from an OpenShift
+// release image.
+func RHCOSBuild(ctx context.Context, releaseImagePullSpec string, pullSecret []byte) (string, error) {
+	imageReferencesBytes, err := imageReferences(ctx, releaseImagePullSpec, pullSecret)
+	if err != nil {
+		return "", err
+	}
+
+	pullSpec, err := osContentPullSpec(imageReferencesBytes)
+	if err != nil {
+		return "", err
+	}
+
+	imageSource, image, err := pull(ctx, pullSpec, pullSecret)
+	if err != nil {
+		return "", err
+	}
+	defer imageSource.Close()
+
+	inspect, err := image.Inspect(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	version, ok := inspect.Labels["version"]
+	if !ok {
+		return "", errors.Errorf("%s has no version annotation", pullSpec)
+	}
+
+	return version, nil
+}
+
+// pull pulls an image.  If the pull secret includes credentials for
+// the pull-spec authority, those are used to authenticate requests.
+// The signature on the returned image, if any, is validated agaist
+// the signature policy from /etc/containers/policy.json.
+func pull(ctx context.Context, pullSpec string, pullSecret []byte) (types.ImageSource, types.Image, error) {
+	named, err := reference.ParseNamed(pullSpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	reference, err := docker.NewReference(named)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sys := &types.SystemContext{}
+	err = addPullSecret(sys, pullSecret, named)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	imageSource, err := reference.NewImageSource(ctx, sys)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() {
+		if err != nil {
+			imageSource.Close()
+		}
+	}()
+
+	unparsed := image.UnparsedInstance(imageSource, nil)
+
+	policy, err := signature.DefaultPolicy(sys)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if allowed, err := policyContext.IsRunningImageAllowed(ctx, unparsed); !allowed || err != nil {
+		return nil, nil, errors.Wrapf(err, "%s rejected", pullSpec)
+	}
+
+	img, err := image.FromUnparsedImage(ctx, sys, unparsed)
+	return imageSource, img, err
+}
+
+func imageReferences(ctx context.Context, releaseImagePullSpec string, pullSecret []byte) ([]byte, error) {
+	imageSource, image, err := pull(ctx, releaseImagePullSpec, pullSecret)
+	if err != nil {
+		return nil, err
+	}
+	defer imageSource.Close()
+
+	layers := image.LayerInfos()
+	for i := len(layers) - 1; i >= 0; i-- {
+		layer := layers[i]
+		blob, _, err := imageSource.GetBlob(ctx, layer, blobinfocache.NoCache)
```

Comment:
If this is expected to enforce/verify signatures (as the code does by calling `IsRunningImageAllowed` in `pull`, this code must enforce that the `blob` stream matches the digest in `layer.Digest`; otherwise the registry could substitute a malicious blob not matching the manifest.

(`IsRunningImageAllowed` can’t do that for you: it would have to download the blobs, verify them, and cache them somewhere, so that later calls to `GetBlob` can read the cached version to ensure it has not been changed; the layering/responsibilities are allocated the opposite way, which allows the caller to stream the data directly with no storage requirements, as long as it handles the verification.

`GetBlob` also does not do digest verification; without, again, requiring extra storage space to cache the data, it could in principle return an error at EOF, but that would only help consumers that don’t consume the data as being streamed, just the way this code does. Also, `GetBlob` is implemented in all the different transports, and having every one of them copy&paste code to verify digests would be hard to maintain and enforce; centralizing this in the consumer, typically `c/image/copy.Image` is more practical.)